### ISSUE TITLE
Type-mismatch error boxed Trait

### ISSUE THREAD

**Issue by villem** [rust-analyzer/rust-analyzer#11818]
Hi,

The latest nightly 3-25-2022 ( 0.3.989) on Mac vscode gives me this error. Cargo build this is ok. Also the RA build 0.3.986 don't 
give this error. rustc -V
rustc 1.58.0 (02072b482 2022-01-11)

[{
	"resource": "/Users/ville.mattila/src/smc_umbrella/backend/src/manager/policy.rs",
	"owner": "rustc",
	"code": {
		"value": "type-mismatch",
		"target": {
			"$mid": 1,
			"external": "https://rust-analyzer.github.io/manual.html#type-mismatch",
			"path": "/manual.html",
			"scheme": "https",
			"authority": "rust-analyzer.github.io",
			"fragment": "type-mismatch"
		}
	},
	"severity": 8,
	"message": "expected Box<dyn Foo<FooItem = <B as Bar>::BarItem>, Global>, found Box<F, Global>",
	"source": "rust-analyzer",
	"startLineNumber": 33,
	"startColumn": 26,
	"endLineNumber": 33,
	"endColumn": 39
}]

Here is source code. I grant this may be most correct rust, but it works for me. You can give suggestion better code too :-) ,
but RA should not give error, where rustc is happy.

```rust

trait Foo {
    type FooItem;
}

trait Bar {
    type BarItem: Foo;
    fn add_foo(&self, bar: Box<dyn Foo<FooItem=Self::BarItem>>) {

    }
}

struct Baz<B> {
    bar: B,
}

impl<B:Bar>  Baz<B> {
    fn add_rule<F: 'static>(&mut self, mut foo: F)
    where F: Foo<FooItem = <B as Bar>::BarItem> {
        self.bar.add_foo(Box::new(foo)); // <<-- THIS line give RA Error now
    }
}


```

**Comment by flodiebold** [comment]
This will be fixed by https://github.com/rust-lang/chalk/pull/754 once that gets merged. Also, https://github.com/rust-analyzer/rust-analyzer/pull/11809 will hide the diagnostic unless you explicitly enable experimental errors :slightly_smiling_face:

**Comment by villem** [comment]
Ok thanks!
